pkg/cafs/reader.go

```diff
@@ -201,115 +273,332 @@ func (r *chunkReader) WriteTo(writer io.Writer) (n int64, err error) {
 
 }
 
-func calculateKeyAndOffset(off int64, leafSize uint32) (index int64, offset int64) {
-	index = off / int64(leafSize)
+func (r *chunkReader) watchPrefetched(wg *sync.WaitGroup, fetchC <-chan fetch) {
+	defer wg.Done()
+
+	// collect prefetched leaf index
+	for f := range fetchC {
+		r.fetchingLatch.Lock()
+		r.fetching[f.index] = f
+		r.fetchingLatch.Unlock()
+	}
+}
+
+func (r *chunkReader) doPrefetch(index, initiator int, logger *zap.Logger) (LeafBuffer, bool, error) {
+	// doPrefetch is in charge of launching go routines to fetch-ahead of the current leaf whenever appropriate:
+	// * fetching ahead is throttled by some parameter (default is 1 ahead)
+	// * we request leaves ahead only once, unless they have been wiped from buffer cache (reduce wasted work)
+	// * we cannot totally rule out wasted work because the buffer cache may not be able to retain all buffers
+	//
+	// A way to assert that there is no more wasted work than allowed by memory constraint:
+	// $ DEBUG_TEST=1 go test -v -run Reader_All|grep -E '(wasted work)|(has read leaf)'|jq -c '{key: .key,msg: .msg}'|sort
+
+	r.fetchingLatch.Lock()
+	defer r.fetchingLatch.Unlock()
+
+	if _, beingFetched := r.fetching[index+1]; !beingFetched {
```

Comment:
Why not just have a hash map of the keys in flight, start a goroutine that waits for the key to be removed from the map and be present in the cache. If we implement a different prefetch scheme the logic will still hold true.

Reply:
I thought about this, but came to recursive decision making about prefetching, reassessing the current situation every time the process requests a new leaf.

Indeed, this is a moving target: new keys to fetch ahead are discovered as the reading goes: every time a leaf is cached, I jump next, and assess the new next fetch-ahead list.

Moreover, the cache cannot guarantee to hold buffers fetched ahead: sometimes (a few as I could, and only when we have much more readers than available cache entries) the prefetch work is wasted (must more uncommon event than in the previous implementation, though) and the leaft must be read again.

Reply:
so basically, the doPrefetch is here to:
i) decide whether to prefetch or not
ii) manage the current list of leafs being prefetched
iii) use the buffer cache as the referee on whether to consider work done or declare it wasted

Reply:
follow-up note: open discussion thread